feat(skorch): add an inherited class from skorch.NeuralNet that is compatible with PyTorch Frame

34j opened the pull request:
Closes #147

~~@MacOS Please continue from here if it helps.
Sorry for being so loud, but this took me a whole day, so I would appreciate it very much if you could make me as a co-author if you used this code.~~

34j committed:
feat(skorch): add prototype of an inherited class from skorch.NeuralNet that is compatible with PyTorch Frame

34j committed:
docs: add tutorial for the last commit

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

34j committed:
fix: patch `skorch.utils.to_tensor()`

MacOS commented:
Sure - on both!

34j committed:
style: format code

34j committed:
feat: fix multiple issues, support sklearn-like datasets and predict()

34j committed:
chore(example): test with regression as well

34j committed:
Merge branch 'master' into feat/skorch-compatible

34j committed:
docs: add changelog

34j committed:
fix(skorch): import annotations from __future__

34j committed:
revert: revert wrong changes

34j commented:
@weihua916 It may not be perfect yet, but would you mind reviewing if this is a good way to implement it?

34j committed:
style(skorch): fix typing

34j committed:
fix(skorch): use `classes` if specified

34j committed:
Merge branch 'master' into feat/skorch-compatible

34j commented:
@weihua916

weihua916 commented on examples/revisiting.py:
```diff
@@ -156,30 +158,87 @@ def test(loader: DataLoader) -> float:
+            lr=args.lr,
+            device=device,
+            verbose=1,
+            # col_to_stype={"C_feature_7": stype.categorical},
```
What's this comment?

weihua916 commented on examples/revisiting.py:
```diff
@@ -156,30 +158,87 @@ def test(loader: DataLoader) -> float:
+            lr=args.lr,
+            device=device,
+            verbose=1,
+            # col_to_stype={"C_feature_7": stype.categorical},
```
What's this comment?

weihua916 reviewed (commented):
Thank you!

Do you mind adding unit test similar to https://github.com/pyg-team/pytorch-frame/blob/master/test/gbdt/test_gbdt.py?

weihua916 commented:
A kind check-in. Is there any progress here?

34j commented:
No progress, sorry

34j committed:
Merge branch 'master' into feat/skorch-compatible

34j committed:
Merge branch 'master' into feat/skorch-compatible

34j committed:
chore: remove comments

34j committed:
chore(skorch): add more comments

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

34j committed:
test: add prototype test

34j commented:
- First, `torch_frame.DataSet` cannot properly handle `DataFrane` with indices not [0,1,...]. `dataset.py` L736 must be modified to `indices = np.where(self.df[self.split_col] == SPLIT_TO_NUM[split])[0]`.
- Second, though I am not aware of this before, `torch_frame` models require `col_stats`, `col_names_dict` in the constructor and unfortunately do not allow modifications after instantiation. Furthermore, they do not follow the standards of scikit-learn as they do not save the constructor parameters, making it impossible to instantiate the module again.

34j committed:
feat: add NeuralNetBinaryClassifierPytorchFrame

34j committed:
test: update test

34j committed:
fix(dataset): fix dataset

34j committed:
feat: allow creating module later

34j committed:
test: add binary test

34j committed:
feat: add sklearn test

34j committed:
style: format code

34j committed:
docs: update docs

34j committed:
chore(deps): add skorch as deps

34j commented:
Almost done

34j committed:
test_skorch.py

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

34j committed:
test_skorch.py

34j committed:
fix: use dict.update instead of dict | dict

34j committed:
fix(dataset): convert indices to list

34j committed:
fix: fix staticmethod usage for < Python 310

https://stackoverflow.com/a/12718272/14819077

34j commented:
@weihua916 Thanks for your patience, I've added tests.

34j committed:
fix: safer patch

34j committed:
fix: do not call twice

34j committed:
fix: copy dataframe before adding columns

zechengz commented on examples/revisiting.py:
```diff
@@ -156,30 +158,85 @@ def test(loader: DataLoader) -> float:
+    print(f'Best Val {metric}: {best_val_metric:.4f}, '
+          f'Best Test {metric}: {best_test_metric:.4f}')
+elif args.framework == "skorch-dataframe":
+    import numpy as np
```
nit: I think we can move these imports to top as we will install these packages for pytorch frame

zechengz commented on examples/tutorial.py:
```diff
@@ -250,17 +252,71 @@ def test(loader: DataLoader) -> float:
+        dim=-1) == test_tensor_frame.y).float().mean()
+    print(f"Test Acc: {test_acc:.4f}")
+elif args.framework == "skorch-dataframe":
+    import pandas as pd
```
Also we can move these to the top

zechengz commented on torch_frame/utils/skorch.py:
```diff
@@ -0,0 +1,313 @@
+def _patch_skorch_support_tenforframe() -> None:
+    old_to_tensor = skorch.utils.to_tensor
+
+    def to_tensor(X, device, accept_sparse=False):
```
Add typing information?

34j replied:
As well, this is not recommended as skorch does not support typing.

zechengz commented on torch_frame/utils/skorch.py:
```diff
@@ -0,0 +1,313 @@
+
+
+# TODO: make it more safe
+def _patch_skorch_support_tenforframe() -> None:
```
Add some docstring?

zechengz commented on torch_frame/utils/skorch.py:
```diff
@@ -0,0 +1,313 @@
+
+# TODO: make it more safe
+def _patch_skorch_support_tenforframe() -> None:
+    old_to_tensor = skorch.utils.to_tensor
```
nit:
```suggestion
    original_to_tensor = skorch.utils.to_tensor
```

zechengz commented on torch_frame/utils/skorch.py:
```diff
@@ -0,0 +1,313 @@
+    def __init__(
+        self,
+        # NeuralNet parameters
+        module,
```
Also add the typing information here?

34j replied:
Since skorch does not support typing, doing this is tedious and not recommended.

zechengz reviewed (commented):
Left some comments. Can @weihua916 or @yiweny helps to take a look? Thanks for your contribution!

34j committed:
docs: add docs to _patch_skorch_support_tenforframe

34j committed:
fix(skorch): wrap with functools.wraps

34j committed:
fix: move imports

34j committed:
chore: do not use NeuralNetClassifierPytorchFrame for regression although it can be used

34j committed:
fix(skorch): add typing only for module

34j committed:
fix: support specifying module as class

34j committed:
docs: add docs

34j committed:
fix: fix dtype

34j committed:
Merge branch 'master' into feat/skorch-compatible

34j committed:
Merge branch 'master' into feat/skorch-compatible

weihua916 commented on examples/revisiting.py:
```diff
@@ -156,30 +161,82 @@ def test(loader: DataLoader) -> float:
+    df_train = pd.concat([train_dataset.df, val_dataset.df])
+    X_train, y_train = df_train.drop(
+        columns=[dataset.target_col, dataset.split_col]), df_train[
+            dataset.target_col]
+    df_test = test_dataset.df
+    X_test, y_test = df_test.drop(
+        columns=[dataset.target_col, dataset.split_col]), df_test[
+            dataset.target_col]
+
+    # use DataFrames with no `split_col` or `target_col`
+    # like normal sklearn datasets from now on
+    if is_classification:
+        net = NeuralNetClassifierPytorchFrame(
+            module=model,
+            criterion=nn.CrossEntropyLoss,
+            max_epochs=args.epochs,
+            lr=args.lr,
+            device=device,
+            verbose=1,
+            batch_size=args.batch_size,
+        )
+    else:
+        net = NeuralNetPytorchFrame(
+            module=model,
+            criterion=nn.MSELoss,
+            max_epochs=args.epochs,
+            lr=args.lr,
+            device=device,
+            verbose=1,
+            batch_size=args.batch_size,
+        )
+    net.fit(X_train, y_train)
+    y_pred = net.predict(X_test)
+
+    if is_classification:
+        test_acc = (y_pred.argmax(-1) == y_test).mean()
+        print(f"Test Acc: {test_acc:.4f}")
+    else:
+        test_rmse = np.sqrt(((y_pred.squeeze() - y_test)**2).mean())
+        print(f"Test RMSE: {test_rmse:.4f}")
```
no need. `tutorial.py` is enough

weihua916 commented on examples/tutorial.py:
```diff
@@ -250,17 +254,68 @@ def test(loader: DataLoader) -> float:
+    test_acc = (torch.Tensor(y_pred).argmax(
+        dim=-1) == test_tensor_frame.y).float().mean()
+    print(f"Test Acc: {test_acc:.4f}")
+elif args.framework == "skorch-dataframe":
+
+    from torch_frame.utils.skorch import NeuralNetClassifierPytorchFrame
+
+    df = dataset.df
+    df_train = pd.concat([train_dataset.df, val_dataset.df])
+    X_train, y_train = df_train.drop(
+        columns=[dataset.target_col, dataset.split_col]), df_train[
+            dataset.target_col]
+    df_test = test_dataset.df
+    X_test, y_test = df_test.drop(
+        columns=[dataset.target_col, dataset.split_col]), df_test[
+            dataset.target_col]
+
+    # use DataFrames with no `split_col` or `target_col`
+    # like normal sklearn datasets from now on
+    net = NeuralNetClassifierPytorchFrame(
+        module=model,
+        criterion=nn.CrossEntropyLoss,
+        max_epochs=args.epochs,
+        lr=args.lr,
+        device=device,
+        verbose=1,
+        col_to_stype={"C_feature_7": stype.categorical},
+        batch_size=args.batch_size,
+    )
+    net.fit(X_train, y_train)
+    y_pred = net.predict(X_test)
+    test_acc = (y_pred.argmax(-1) == y_test).mean()
+    print(f"Test Acc: {test_acc:.4f}")
```
I'd isolate this into another stand-alone file under `examples`

weihua916 commented on test/utils/test_skorch.py:
```diff
@@ -0,0 +1,207 @@
+    )
+    dataset.materialize()
+    train_dataset, val_dataset, test_dataset = dataset.split()
+    # print(dataset.col_stats)
+    # # convert to dataframe
+    # col_to_stype = dataset.col_to_stype
+    # # remove split_col and target_col
+    # col_to_stype = {
+    #     k: v
+    #     for k, v in col_to_stype.items()
+    #     if k not in [dataset.split_col, dataset.target_col]
+    # }
```
remove

34j committed:
test: remove comment

34j committed:
Discard changes to examples/revisiting.py

34j committed:
Discard changes to examples/tutorial.py

34j committed:
Discard changes to README.md

34j committed:
fix: use args instead of kwargs to match typing

34j committed:
feat: add example for sklearn api

34j commented:
Removed all tutorials and added `examples/sklearn_api.py` instead. 
The failing tests are probably due to a pandas version update and are not related to this PR.

weihua916 commented on test/utils/test_skorch.py:
```diff
@@ -0,0 +1,198 @@
+    [
+        [stype.numerical],
+        [stype.categorical],
+        # [stype.text_embedded],
+        # [stype.numerical, stype.numerical, stype.text_embedded],
```
we don't support these stypes?

34j replied:
Currently not supported at this time due to lack of time to understand how to use these dtypes.
However, since it probably only require changes in the arguments of the NeuralNet, it should have little trouble extending it in the future.

weihua916 commented on test/utils/test_skorch.py:
```diff
@@ -0,0 +1,198 @@
+    elif task_type == TaskType.BINARY_CLASSIFICATION:
+        net = NeuralNetBinaryClassifierPytorchFrame(**kwargs, )
+
+    if pass_dataset:
+        net.fit(dataset)
+        _ = net.predict(test_dataset)
+    else:
+        net.fit(X_train, y_train)
+        _ = net.predict(X_test)
```
why don't we take tensor frame? It's also weird to sometimes take dataset and sometimes take data frame.

34j replied:
- The main purpose of this PR is to allow DataFrame to be DIRECTLY fitted.
- Since it is unclear how to create a Dataset from a TensorFrame, and if there is a TensorFrame, there is also a Dataset, which means there is little need to use skorch.
- Instead of "sometimes take dataset and sometimes take data frame", both are tested.

weihua916 replied:
hmm if dataframe is directly fed, it is unclear why we need this feature within pytorch frame.
the whole point of pytorch frame is to materialize data frame into tensor frame, to be processed by pytorch.

34j replied:
It may not match your purpose but my goal is to use advanced neural networks implemented in pytorch_frame in existing sklearn pipeline. 
This PR allows pytorch_frame to be used on top of existing scikit-learn code without having to heavily modify the existing code. Since many people use the sklearn pipeline, especially on Kaggle, it is easy to verify performance changes by changing or assembling the estimator in other people's code to my `NeuralNetPytorchFrame`. I am innovating this as a very win-win. I am convinced that this will be very valuable.

weihua916 reviewed (commented):
Left a quick question.

34j committed:
Merge branch 'master' into feat/skorch-compatible

34j committed:
Merge branch 'master' into feat/skorch-compatible

34j commented:
@weihua916 Would you please consider merging this PR again?

34j committed:
Merge branch 'master' into feat/skorch-compatible

34j commented:
@weihua916 @yiweny @akihironitta Any chance this PR could be merged?

akihironitta reviewed (changes requested):
Great work on this integration! 🚀 Thank you for your effort :)

After reviewing the PR, I'd recommend keeping this integration as an example script rather than merging it into `torch_frame/`. Since this integration may not see widespread or frequent use, maintaining the code (e.g., keeping up with new APIs `skorch` introduces and making sure all features work in PyTorch Frame) could be challenging in the future. IMO, presenting it as an example would make more sense. Alternatively, @34j (and contributors in the linked issue together) could consider creating a separate repository to showcase this integration as a community-driven effort, and we could help highlight it via our docs or social media. (We can always promote the integration into our codebase in the future if it gains traction and the community expresses a need for it.)

Looks like PyTorch Frame doesn't need many changes to support this integration, but I'm happy to provide any extension points when it's necessary.

I'm open to any comments/feedbacks/disussions!

34j commented:
In that case I would consider making another package, but I am very busy right now, so please leave this as it is for a while. Note that in any case one-line change I made (in this PR) needs to be merged for that to work.
